001-collection-backend-dispatching.md

```diff
@@ -0,0 +1,165 @@
+  - Dask should clearly define the necessary API for defining a new backend.
+
+
+## Non-Goals
+
+- This feature does not target type- or hardware-based computation dispatching on the worker. The new dispatching system only applies to collection-API usage on the client process.
+- This feature does not target the conversion of an existing collection class to a different backend library after the initial collection is created. That is, if the backend configuration is set after the collection is already defined, the backend will **not** be moved to the desired library.
+- This feature does not propose a gpu (or RAPIDS) on/off switch. In the future, the backend defaults can be modified to depend on other hardware-preference information stored in the `dask.config` options. However, this feature only calls for backend configuration options at the collection level.
+- This feature does not mean Dask will take responsibility for testing a `cudf`-backed version of `dask.dataframe`. It is still the responsibility of the backend library (i.e. `cudf` and `dask_cudf`) to ensure that the collection and backend libraries are compatible.
+
+
+## Detailed Description
+
+
+### Designating the Backend (`dask.config`)
+
+From the perspective of the typical Dask user, the only visible result of the proposed feature is the addition of new fields in `dask/dask/dask.yaml`/`dask-schema.yaml` (accessiblke from `dask.config`). For each of the targeted collections (Dask-Array and Dask-DataFrame), we propose the addition of "backend", and "backend-options" fields. By default, the "backend" field will be set to "numpy" and "pandas" for Dask-Array and Dask-DataFrame, respectively. However, as shown in the code snippet below, this field can be changed with the existing `dask.config` interface to specify an alternative backend library.
+
+
+```python
+import dask
+
+with dask.config.set({"dataframe.backend": "cudf"}):
+    # Produce a cudf-backed collection
+    ddf = dask.dataframe.read_parquet("./tmpdir")
+
+with dask.config.set({"array.backend": "cupy"}):
+    # Produce a cupy-backed collection
+    darr = dask.array.ones(10, chunks=(5,))
+```
+
+Since it is unlikely that an alternative backend will support all numpy- or pandas-based data-creation functions available in the collection API, we also propose a "backend-options" field containing "allow-fallback" and "warn-fallback" subfields. When the "allow-fallback" field is set to `True` (default), then the backend's designated fallback class will be used to perform IO, and the result will be moved from the fallback backend. The user should also have the option to enable or disable warnings when this fallback behavior occurs:
+
+
+```python
+options = {
+    # Allow user to specify if a fallback backend library
+    # should be used, and if a warning should be raised when
+    # this occurs:
+    "allow-fallback" : True,
+    "warn-fallback" : True,
+}
+
+dask.config.set({"dataframe.backend-options": options})
```

Comment:
>Do you think it would be confusing to allow a string OR a dict

I think that makes a lot of sense from the perspective of the user, but wouldn't that prevent us from defining/documenting anything below "backend" in the yaml config file?  Wouldn't we need to require the user to always set the backend with `"dataframe.backend.libarary"`? (note that I may not be understanding how the config file works, and adding the extra "libarary" string is probably fine anyway).

Reply:
yeah it might not work very well. I guess I was imagining that `backend` would just be an alias for `backend.library` but maybe it's not a good idea to special-case things like that.

Reply:
I changed `"<collection>.backend"` to `"<collection>.backend.library"`, and moved "allow-fallback" and "warn-fallback" under "backend".  This is not exactly what was suggested, but seems better than what I initially had.